components/raftstore/src/store/fsm/peer.rs

```diff
@@ -751,46 +751,45 @@ where
         if !self.fsm.peer.is_force_leader() {
             error!(
                 "Unsafe recovery, demoting failed voters failed, since this peer is not forced leader";
-                "region" => ?self.region(),
+                "region_id" => self.region().get_id(),
+                "peer_id" => self.fsm.peer.peer.get_id(),
             );
             return;
         }
 
         if self.fsm.peer.unsafe_recovery_state.is_some() {
             warn!(
                 "Unsafe recovery, demote failed voters has already been initiated";
-                "region" => ?self.region(),
+                "region_id" => self.region().get_id(),
+                "peer_id" => self.fsm.peer.peer.get_id(),
             );
             return;
         }
 
         if self.fsm.peer.in_joint_state() {
             info!(
                 "Unsafe recovery, already in joint state, exit first";
-                "region" => ?self.region());
-            let proposed = Arc::new(Mutex::new(false));
-            let proposed_clone = proposed.clone();
-            let callback = Callback::<EK::Snapshot>::write_ext(
-                Box::new(|resp| {
-                    if resp.response.get_header().has_error() {
-                        error!(
-                            "Unsafe recovery, fail to exit residual joint state";
-                            "err" => ?resp.response.get_header().get_error(),
-                        );
-                    }
-                }),
-                Some(Box::new(move || {
-                    *proposed_clone.lock().unwrap() = true;
-                })),
-                None,
+                "region_id" => self.region().get_id(),
+                "peer_id" => self.fsm.peer.peer.get_id(),
             );
+            let failed = Arc::new(Mutex::new(false));
+            let failed_clone = failed.clone();
+            let callback = Callback::<EK::Snapshot>::write(Box::new(move |resp| {
+                if resp.response.get_header().has_error() {
+                    *failed_clone.lock().unwrap() = true;
+                    error!(
+                        "Unsafe recovery, fail to exit residual joint state";
+                        "err" => ?resp.response.get_header().get_error(),
+                    );
+                }
+            }));
             self.propose_raft_command_internal(
                 exit_joint_request(self.region(), &self.fsm.peer.peer),
                 callback,
                 DiskFullOpt::AllowedOnAlmostFull,
             );
 
-            if *proposed.lock().unwrap() {
+            if !*failed.lock().unwrap() {
```

Comment:
the callback is called after applied, the failed won't be filled right after calling propose. You can just check if the `last index` is changed before and after identifying the proposal is accepted.

Reply:
Oh, the callback is called when encountering any error. Seem it's okay. Please add test to cover this